src/wallet/bdk.rs

```diff
@@ -1363,6 +1451,142 @@ impl Wallet {
+}
+
+impl Wallet {
+    fn create_wallet_descriptors(&self) -> Result<HashMap<KeychainKind, String>, anyhow::Error> {
+        // create master_key from seed
+
+        let secp = Secp256k1::new();
+
+        // create Xpriv from HARDENDED_DERIVATION path(till account no):
+        let wallet_xpriv = self
+            .store
+            .master_key
+            .derive_priv(
+                &secp,
+                &DerivationPath::from_str(HARDENDED_DERIVATION).unwrap(),
+            )
+            .unwrap();
+
+        // create descriptors for each keychain -> get hashmap.
+        // Note:
+        // created descriptor will contain xpriv not xpub
+
+        let mut kind_descriptor_map = HashMap::new();
+        //  let descriptors =
+        for keychain in KeychainKind::iterator() {
+            let descriptor_without_checksum = match *keychain {
+                KeychainKind::External | KeychainKind::Internal => {
+                    format!("wpkh({}/{}/*)", wallet_xpriv, keychain.index_num())
+                }
+                KeychainKind::Fidelity | KeychainKind::SwapCoin | KeychainKind::Contract => todo!(),
+            };
+            let descriptor = format!(
+                "{}#{}",
+                descriptor_without_checksum,
+                calc_checksum(&descriptor_without_checksum).unwrap()
+            );
+            // TODO: will always return None
+            kind_descriptor_map.insert(*keychain, descriptor);
+        }
+
+        Ok(kind_descriptor_map)
+    }
+}
+
+//_____________________________________WALLET/SIGNER______________________________________________
+impl Wallet {
+    fn create_signers<E>(
+        index: &mut KeychainTxOutIndex<KeychainKind>,
+        secp: &Secp256k1<All>,
+        kind_descriptor_map: HashMap<KeychainKind, E>,
+        network: Network,
+    ) -> Result<HashMap<KeychainKind, Arc<SignersContainer>>, DescriptorError>
+    where
+        E: IntoWalletDescriptor,
+    {
+        let mut signers_map: HashMap<KeychainKind, Arc<SignersContainer>> = HashMap::new();
+
+        for (kind, descriptor) in kind_descriptor_map.into_iter() {
+            let (descriptor, keymap) = into_wallet_descriptor_checked(descriptor, secp, network)?;
+            let signers = Arc::new(SignersContainer::build(keymap, &descriptor, &secp));
+            let _ = index.insert_descriptor(kind, descriptor);
+            signers_map.insert(kind, signers);
+        }
+
+        Ok(signers_map)
+    }
+
+    pub fn get_signers(&self, keychain: KeychainKind) -> Arc<SignersContainer> {
+        Arc::clone(self.signers.get(&keychain).expect("must not fail"))
+    }
+
+    pub fn add_signer(
+        &mut self,
+        keychain: KeychainKind,
+        ordering: SignerOrdering,
+        signer: Arc<dyn TransactionSigner>,
+    ) {
+        let signer_container = Arc::make_mut(self.signers.get_mut(&keychain).expect("must exist"));
+
+        signer_container.add_external(signer.id(&self.secp), ordering, signer);
+    }
```

Comment:
I don't think we will need this. The wallet will always start with all the signers it needs.

Reply:
> instead we can write its code wherever we want to add signer.

Exactly.

Reply:
ACK. It's the same thing.

Just write the whole thing first in a single function. Break them later. 

Don't create API before writing the code. Write the code and see how the API needs to be created.